### DISCUSSION TITLE
Difference between cpu and itimer mode

### DISCUSSION THREAD

**Discussion by infa-jmashalk** [async-profiler/async-profiler#687] [Q&A] [answered]
Hi, can someone explain the difference between the itimer and cpu mode? I understand itimer mode can be run when enabling perf_events is not feasible, hence looking to use this option in our production where there are strict controls on linux filesystem and root user access. 
Hence want to know what are the pitfalls of using this mode, or what are the additional details we get using the cpu mode.

**Comment by apangin** [comment, marked as answer]
Please see #272.

In short, `cpu` mode (based on perf_events) provides kernel stack traces and is more accurate overall.

**Reply by infa-jmashalk** [reply]
ok, thanks. are there any other restrictions? are these options allowed:
-G1CollectedHeap::humongous_obj_allocate
--live - retain allocation samples with live objects only (object that have not been collected by the end of profiling session). Useful for finding Java heap memory leaks.

--lock N - lock profiling threshold in nanoseconds (or other units). In lock profiling mode, record contended locks that the JVM has waited for longer than the specified duration.
Not getting proper outputs in flamegraph since kernel events have been turned off.

**Reply by apangin** [reply]
I don't see how this relates to the original question.

Native function profiling (in particular, `-e G1CollectedHeap::humongous_obj_allocate`) is a separate mode, not related to cpu/itimer. It requires access to `perf_events`.

`--live` is specific to allocaton profiling (`alloc`), it works in JDK 11+.
`--lock` is lock contention profiling - yet another mode, not related to cpu/itimer.
Neither `alloc` nor `lock` relies on kernel capabilities.
